draft-ietf-tls-esni.md

```diff
@@ -444,10 +446,40 @@ values, ClientHelloInner MUST also contain:
  - an "echo_nonce" extension
  - TLS padding {{!RFC7685}}
 
-Padding SHOULD be P = L - D bytes, where
-
-- L = ECHOConfig.maximum_name_length, rounded up to the nearest multiple of 16
-- D = len(dns_name), where dns_name is the DNS name in the ClientHelloInner "server_name" extension
+Variations in the length of the ciphertext version of the ClientHelloInner
+could defeat the purpose of ECHO, if those expose the chosen server_name field
+in the ClientHelloInner. The padding ClientHello extension, if well used, can
+ensure that length information does not expose what is contained in the
+ClientHelloInner. Clients SHOULD add padding to the ClientHelloInner to meet this
+requirement.
+
+Given that extensions could be defined in the future that reveal sensitive information
+through their length, we cannot describe all the ways in which length information
+could expose sensitive content. Implementers ought therefore be aware that they
+might have to change their padding scheme as the set of supported extensions
+changes. 
+
+For each field in the inner ClientHello, clients need to determine how much to
+pad given the semantics of that field.  For example, if a client can propose
+various ALPN values, it could add padding to round up to the longest of those.
+
+For most fields in a ClientHello this can be determined without server help.
+For the server_name, however, if ECHOConfig.maximum_name_length is longer than
+the actual server_name then clients SHOULD add padding to make up that
+difference. If the maximum_name_length is zero or less than the length of the
+actual server_name then round the server_name up to a multiple of 32 octets and
+randomly add another 32 octets 50% of the time and then include that amount
+of additional padding.
```

Comment:
```suggestion
However, the "server_name" extension could benefit from server input (ECHOConfig.maximum_name_length). 
Clients SHOULD compute the padding for this extension as follows:

1. If ECHOConfig.maximum_name_length is longer than the actual server_name then 
clients SHOULD add padding to make up that difference.
2. Otherwise, if ECHOConfig.maximum_name_length is zero or less than the length 
of the actual server_name then round the server_name up to a multiple of 32 octets 
and randomly add another 32 octets 50% of the time and then include that amount
of additional padding.
```